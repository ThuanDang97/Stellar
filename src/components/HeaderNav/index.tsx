import { FC, memo, useMemo, useState, useCallback, useEffect } from 'react'
import Link from 'next/link'
import { useRouter } from 'next/router'

// Props type
import {
  HeaderNavListProps,
  HeaderNavItemProps,
} from '@self-types/components/HeaderNav.props'

// Themes
import metrics from '@themes/Metrics'
import colors from '@themes/Colors'

// Utils
import { updateActiveNavItem } from '@utils/index'

// Components
import {
  ListItemStyled,
  ListStyled,
} from '@components/styled-components/ListStyled'
import { LinkStyled } from '@components/styled-components/LinkStyled'

const HeaderNav: FC<HeaderNavListProps> = ({ headerNavList, isMobile }) => {
  const router = useRouter()
  const { pathname } = router || {}

  const [currentPage, setCurrentPage] = useState(pathname)

  useEffect(() => {
    setCurrentPage(pathname)
  }, [pathname])

  // Handle change current page
  const handleChangeCurrentPage = useCallback((url: string) => {
    setCurrentPage(url)
  }, [])

  const listNavItemUpdate: HeaderNavItemProps[] = useMemo(
    () => updateActiveNavItem(currentPage, headerNavList),
    [currentPage, headerNavList],
  )

  return (
    <ListStyled flexDirection="row">
      {listNavItemUpdate.map((item: HeaderNavItemProps) => {
        const handleClickItem = () => handleChangeCurrentPage(item.name)
        return (
          <ListItemStyled
            key={item.id}
            bgColorHover="none"
            onClick={handleClickItem}
            data-testid={`item${item.id}`}
          >
            <Link href={item.url} passHref>
              <LinkStyled
                className={item.isActive ? 'active' : ''}
                color={colors.white}
                mLeft={isMobile ? 0 : metrics.dimensions.sm}
                mRight={metrics.dimensions.sm}
                data-testid={`${item.url}-${item.id}`}
              >
                {item.name}
              </LinkStyled>
            </Link>
          </ListItemStyled>
        )
      })}
    </ListStyled>
  )
}

export default memo(HeaderNav)
