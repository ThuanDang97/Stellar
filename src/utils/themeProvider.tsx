import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
