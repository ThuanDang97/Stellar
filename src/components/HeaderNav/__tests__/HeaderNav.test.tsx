import '@testing-library/jest-dom'
import React from 'react'
import { fireEvent, screen } from '@testing-library/react'

// Components
import HeaderNav from '@components/HeaderNav/index'

// Constants
import { HEADER_NAVIGATION } from '@constants/index'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

describe('HeaderNav component', () => {
  it('renders HeaderNav correctly', () => {
    const { asFragment } = renderWithTheme(
      <HeaderNav headerNavList={HEADER_NAVIGATION} />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should active text when click in text ', () => {
    renderWithTheme(<HeaderNav headerNavList={HEADER_NAVIGATION} />)
    const item = screen.getByTestId('item1')
    fireEvent.click(item)
    expect(screen.getByTestId('/help-1')).toHaveStyle(`
      color: rgb(255, 255, 255)`)
  })
})
