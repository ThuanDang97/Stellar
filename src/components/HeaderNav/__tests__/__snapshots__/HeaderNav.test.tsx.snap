// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderNav component renders HeaderNav correctly 1`] = `
<DocumentFragment>
  <ul
    class="ListStyled-sc-t03hvr-0 cYiWEp"
  >
    <li
      class="ListStyled__ListItemStyled-sc-t03hvr-1 GyEBw"
      data-testid="item1"
    >
      <a
        class="LinkStyled-sc-626068-0 domzJV"
        color="#FFF"
        data-testid="/help-1"
        href="/help"
      >
        Help
      </a>
    </li>
    <li
      class="ListStyled__ListItemStyled-sc-t03hvr-1 GyEBw"
      data-testid="item2"
    >
      <a
        class="LinkStyled-sc-626068-0 domzJV"
        color="#FFF"
        data-testid="/feedback-2"
        href="/feedback"
      >
        Feedback
      </a>
    </li>
    <li
      class="ListStyled__ListItemStyled-sc-t03hvr-1 GyEBw"
      data-testid="item3"
    >
      <a
        class="LinkStyled-sc-626068-0 domzJV"
        color="#FFF"
        data-testid="/refer-a-friend-3"
        href="/refer-a-friend"
      >
        Refer A Friend
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
